subprojects/groovy-json/src/main/java/groovy/json/JsonOutput.java

```diff
@@ -601,4 +280,966 @@ public String toString() {
+        }
+
+        /**
+         * Excludes from the output any fields that match the specified names.
+         *
+         * @param fieldNames collection of names to exclude from the output
+         * @return a reference to this {@code Options} instance
+         */
+        public Options excludeFieldsByName(Iterable<? extends CharSequence> fieldNames) {
+            for (CharSequence cs : fieldNames) {
+                if (cs != null) {
+                    excludedFieldNames.add(cs.toString());
+                }
+            }
+            return this;
+        }
+
+        /**
+         * Excludes from the output any fields whose type is the same or is
+         * assignable to any of the given types.
+         *
+         * @param types excluded from the output
+         * @return a reference to this {@code Options} instance
+         */
+        public Options excludeFieldsByType(Class<?>... types) {
+            return excludeFieldsByType(Arrays.asList(types));
+        }
+
+        /**
+         * Excludes from the output any fields whose type is the same or is
+         * assignable to any of the given types.
+         *
+         * @param types collection of types to exclude from the output
+         * @return a reference to this {@code Options} instance
+         */
+        public Options excludeFieldsByType(Iterable<Class<?>> types) {
+            for (Class<?> c : types) {
+                if (c != null) {
+                    excludedFieldTypes.add(c);
+                }
+            }
+            return this;
+        }
+
+        /**
+         * Creates a {@link JsonOutput.Generator} that is based on the current options.
+         *
+         * @return a fully configured {@link JsonOutput.Generator}
+         */
+        public Generator createGenerator() {
+            return new Generator(this);
+        }
+    }
+
+    /**
+     * A JsonOutput Generator that can be configured with various {@link JsonOutput.Options}.
+     * If the default options are sufficient consider using the static {@code JsonOutput.toJson}
+     * methods.
+     *
+     * @see JsonOutput#options()
+     * @see Options#createGenerator()
+     * @since 2.5
+     */
+    public static class Generator {
+
+        private final boolean excludeNulls;
+        private final boolean disableUnicodeEscaping;
+        private final String dateFormat;
+        private final Locale dateLocale;
+        private final TimeZone timezone;
+
+        private final Set<Converter> converters = new LinkedHashSet<Converter>();
+
+        private final Set<String> excludedFieldNames = new HashSet<String>();
+
+        private final Set<Class<?>> excludedFieldTypes = new HashSet<Class<?>>();
+
+        private final String nullValue;
+
+        private final boolean hasConverters;
+        private final boolean hasExcludedFieldNames;
+        private final boolean hasExcludedFieldTypes;
+
+        private Generator(Options options) {
```

Comment:
It does feel like that. I do think we should consider extensibility that way if users face an issue with customization it can easily be resolved. Another method I would like protected is `getObjectProperties(Object object)` since frameworks may need to customize that if they add additional properties to objects that are internal. We face this problem a lot in Grails.